src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/TestHooks/ValueTaskExtensions.cs

```diff
@@ -0,0 +1,255 @@
+        {
+            /// <summary>
+            /// The task.
+            /// </summary>
+            private readonly ValueTask<TResult> _task;
+
+            /// <summary>
+            /// A value indicating whether the continuation should be scheduled on the current sync context.
+            /// </summary>
+            private readonly bool _captureContext;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="NoThrowValueTaskAwaiter{TResult}"/> struct.
+            /// </summary>
+            /// <param name="task">The task.</param>
+            /// <param name="captureContext">if set to <c>true</c> [capture context].</param>
+            public NoThrowValueTaskAwaiter(ValueTask<TResult> task, bool captureContext)
+            {
+                Contract.ThrowIfNull(task, nameof(task));
+                _task = task;
+                _captureContext = captureContext;
+            }
+
+            /// <summary>
+            /// Gets a value indicating whether the task has completed.
+            /// </summary>
+            public bool IsCompleted
+            {
+                get { return _task.IsCompleted; }
+            }
+
+            /// <summary>
+            /// Schedules a delegate for execution at the conclusion of a task's execution.
+            /// </summary>
+            /// <param name="continuation">The action.</param>
+            public void OnCompleted(Action continuation)
+            {
+                _task.ConfigureAwait(_captureContext).GetAwaiter().OnCompleted(continuation);
+            }
+
+            /// <summary>
+            /// Schedules a delegate for execution at the conclusion of a task's execution
+            /// without capturing the ExecutionContext.
+            /// </summary>
+            /// <param name="continuation">The action.</param>
+            public void UnsafeOnCompleted(Action continuation)
+            {
+                _task.ConfigureAwait(_captureContext).GetAwaiter().UnsafeOnCompleted(continuation);
+            }
+
+            /// <summary>
+            /// Does nothing.
+            /// </summary>
+            public (TaskStatus Status, ExceptionDispatchInfo? Exception, TResult? Result) GetResult()
+            {
+                var preserved = _task.Preserve();
+                if (preserved.IsCompletedSuccessfully)
+                    return (TaskStatus.RanToCompletion, null, preserved.Result);
+                else if (preserved.IsCanceled)
+                    return (TaskStatus.Canceled, ExceptionDispatchInfo.Capture(preserved.AsTask().Exception!.InnerException!), default);
+                else if (preserved.IsFaulted)
+                    return (TaskStatus.Faulted, ExceptionDispatchInfo.Capture(preserved.AsTask().Exception!.InnerException!), default);
+                else
+                    throw new InvalidOperationException($"Cannot call {nameof(GetResult)} prior to completion.");
+            }
+        }
+    }
+}
```

Comment:
very worried about the difficulty validating that logic like this is correct.  I would prefer we just wait on the compiler doing this work for us so we can get the battle testing of that.

Reply:
I'd be happy to have @stephentoub review this. I would consider it a candidate for inclusion in vs-threading along with the `NoThrowAwaitable` method they already have there for `Task<TResult>`.

Reply:
Personally, I'm not a fan of changing the await to return a tuple of the status, exception, and result and would hope we wouldn't promote that. We're adding support for no-throw awaiting in .NET 8 and, after many go-arounds on design and discussing doing this exact thing and various variations, are very explicitly only doing so for Task and not ValueTask. If you really need to do this with ValueTask, you have various options, like
```C#
vt = vt.Preserve();
await vt.NoThrow(); // your simple awaiter that just nops GetResult
... examine vt
```
and it doesn't require inventing a complicated thing or using an unfamiliar pattern.

Reply:
⏱️ Works for me. I'll rewrite to this pattern.

Reply:
➡️ This has now been rewritten.

Reply:
➡️ This code will be rewritten to use https://github.com/microsoft/vs-threading/pull/1193 after 17.7 Preview 2 ships.

Reply:
was this done?